src/client/oid.rs

```diff
@@ -0,0 +1,216 @@
+use libc;
+
+use client::Result;
+use client::Error::{ArgumentError, DefaultError};
+
+use crypto::digest::Digest;
+use crypto::md5::Md5;
+
+use byteorder::{ByteOrder, BigEndian, LittleEndian};
+use rand::{Rng, OsRng};
+use rustc_serialize::hex::FromHex;
+use time;
+
+use std::sync::atomic::{AtomicUsize, Ordering, ATOMIC_USIZE_INIT};
+
+static MAX_U24: usize = 0xFFFFFF;
+static OID_COUNTER: AtomicUsize = ATOMIC_USIZE_INIT;
+static mut machine_bytes: [u8; 3] = [0; 3];
+
+extern {
+    fn gethostname(name: *mut libc::c_char, size: libc::size_t) -> libc::c_int;
+}
+
+/// Generates a new ObjectID, represented in bytes.
+/// See the [docs](http://docs.mongodb.org/manual/reference/object-id/)
+/// for more information.
+pub fn generate() -> Result<[u8; 12]> {
+    let timestamp = gen_timestamp();
+    let machine_id = try!(gen_machine_id());
+    let process_id = gen_pid();
+    let counter = try!(gen_count());
+
+    let mut buf: [u8; 12] = [0; 12];
+    for i in 0..4 { buf[i] = timestamp[i]; }
+    for i in 0..3 { buf[4+i] = machine_id[i]; }
+    for i in 0..2 { buf[7+i] = process_id[i]; }
+    for i in 0..3 { buf[9+i] = counter[i]; }
+
+    Ok(buf)
+}
+
+/// Creates an ObjectID using a 12-byte (24-char) hexadecimal string.
+pub fn with_string(s: &str) -> Result<[u8; 12]> {
+    let bytes = try!(s.from_hex());
+    if bytes.len() != 12 {
+        Err(ArgumentError("Provided string must be a 12-byte hexadecimal string.".to_owned()))
+    } else {
+        let mut byte_array: [u8; 12] = [0; 12];
+        for i in 0..12 {
+            byte_array[i] = bytes[i];
+        }
+        Ok(byte_array)
+    }
+}
+
+/// Creates a dummy ObjectId with a specific generation time.
+/// This method should only be used to do range queries on a field
+/// containing ObjectId instances.
+pub fn with_timestamp(time: u32) -> [u8; 12] {
+    let mut buf: [u8; 12] = [0; 12];
+    BigEndian::write_u32(&mut buf,time);
+    buf
+}
+
+/// Retrieves the timestamp (seconds since epoch) from an ObjectId.
+pub fn get_timestamp(oid: [u8; 12]) -> u32 {
+    BigEndian::read_u32(&oid)
+}
+
+/// Retrieves the machine id associated with an ObjectId.
+pub fn get_machine_id(oid: [u8; 12]) -> u32 {
+    let mut buf: [u8; 4] = [0; 4];
+    for i in 0..3 {
+        buf[i] = oid[4+i];
+    }
+    LittleEndian::read_u32(&buf)
+}
+
+/// Retrieves the process id associated with an ObjectId.
+pub fn get_pid(oid: [u8; 12]) -> u16 {
+    LittleEndian::read_u16(&oid[7..])
```

Comment:
I don't want to see "magic" number around. Define special numbers as constant. For example: `kTimestampOffset`, `kInstanceUniqueOffset`, and `kIncrementOffset` instantiated here: thttps://github.com/mongodb/mongo/blob/master/src/mongo/bson/oid.cpp#L47